client/vrf/src/lib.rs

```diff
@@ -0,0 +1,89 @@
+// Copyright 2019-2022 PureStake Inc.
+// This file is part of Moonbeam.
+
+// Moonbeam is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Moonbeam is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Moonbeam.  If not, see <http://www.gnu.org/licenses/>.
+
+//! VRF client primitives for client-side verification
+
+use nimbus_primitives::NimbusId;
+use session_keys_primitives::{make_transcript, make_transcript_data, PreDigest, VrfApi, VrfId};
+use sp_application_crypto::{AppKey, ByteArray};
+use sp_consensus_babe::Slot;
+use sp_consensus_vrf::schnorrkel::{PublicKey, VRFOutput, VRFProof};
+use sp_core::H256;
+use sp_keystore::{SyncCryptoStore, SyncCryptoStorePtr};
+
+/// Uses the runtime API to get the VRF inputs and sign them with the VRF key that
+/// corresponds to the authoring NimbusId
+pub fn vrf_pre_digest<B, C>(
+	client: &C,
+	keystore: &SyncCryptoStorePtr,
+	nimbus_id: NimbusId,
+	parent: H256,
+) -> Option<sp_runtime::generic::DigestItem>
+where
+	B: sp_runtime::traits::Block<Hash = sp_core::H256>,
+	C: sp_api::ProvideRuntimeApi<B>,
+	C::Api: VrfApi<B>,
+{
+	let at = sp_api::BlockId::Hash(parent);
+	let relay_slot_number: Slot = client
+		.runtime_api()
+		.get_relay_slot_number(&at)
+		.expect("api error");
+	let relay_storage_root: H256 = client
+		.runtime_api()
+		.get_relay_storage_root(&at)
+		.expect("api error");
```

Comment:
More descriptive error messages would help here. Also, making them unique will help identify problems if we see this in a log.